source/common/router/router.cc

```diff
@@ -1329,6 +1329,25 @@ bool Filter::maybeRetryReset(Http::StreamResetReason reset_reason,
                           ? RetryState::Http3Used::Yes
                           : RetryState::Http3Used::No;
   }
+
+  auto upstream_request_started = false;
+  // TODO: Is this assertion worth the risk?
+  ASSERT(upstream_requests_.size() <= 2);
```

Comment:
I'd say no - if this changes to be large whoever changes it should handle checking for scale.